keys.go

```diff
@@ -50,116 +74,240 @@ func findKey(session *pkcs11Session, id []byte, label []byte, keyclass *uint, ke
 	if label != nil {
 		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_LABEL, label))
 	}
-	if err = session.ctx.FindObjectsInit(session.handle, template); err != nil {
-		return nil, err
-	}
-	defer func() {
-		finalErr := session.ctx.FindObjectsFinal(session.handle)
-		if err == nil {
-			err = finalErr
-		}
-	}()
-	if handles, _, err = session.ctx.FindObjects(session.handle, 1); err != nil {
+
+	if handles, err = findKeysWithAttributes(session, template); err != nil {
 		return nil, err
 	}
+
+	return handles, nil
+}
+
+// Find a key object.  For asymmetric keys this only finds one half so
+// callers will call it twice. Returns nil if the key does not exist on the token.
+func findKey(session *pkcs11Session, id []byte, label []byte, keyclass *uint, keytype *uint) (obj *pkcs11.ObjectHandle, err error) {
+	handles, err := findKeys(session, id, label, keyclass, keytype)
+
 	if len(handles) == 0 {
 		return nil, nil
 	}
 	return &handles[0], nil
 }
 
+// Takes a handles to the private half of a keypair, locates the public half with the matching CKA_ID and CKA_LABEL
+// values and constructs a keypair object from them both.
+func (c *Context) makeKeyPair(session *pkcs11Session, privHandle *pkcs11.ObjectHandle) (signer Signer, err error) {
+	attributes := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_ID, nil),
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, nil),
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, 0),
+	}
+	if attributes, err = session.ctx.GetAttributeValue(session.handle, *privHandle, attributes); err != nil {
+		return nil, err
+	}
+	id := attributes[0].Value
+	label := attributes[1].Value
+	keyType := bytesToUlong(attributes[2].Value)
+
+	// Ensure the private key actually has a non-empty CKA_ID to match on
+	if id == nil || len(id) == 0 {
+		return nil, errors.New("private key has no CKA_ID")
```

Comment:
We should return a specific error or a nil Signer. That would allow the calling functions to differentiate a bad error (e.g. P11 not working any more).

Currently, the calling function (`FindKeyPairsWithAttributes`) ignores all errors coming from this function. Which doesn't feel like a good plan.

Reply:
This is tricky because I can imagine there would be some cases where you might want it to skip past these keys so it can then find other keys which are valid, like it does for the cases when you're finding multiple keys, rather than failing hard. You wouldn't necessarily want one dodgy key to prevent it from being able to find any other keys.

Reply:
I'm not suggesting we give up finding key pairs because one pair hasn't got a `CKA_ID`, or whatever. If we return a known error, we can handle that differently in the `FindKeyPairsWithAttributes` function:

```go
// inside FindKeyPairsWithAttributes...

for _, privHandle := range privHandles {
  k, err := c.makeKeyPair(session, &privHandle)
  if err == errNoCkaID {
    continue
  }
  if err != nil {
    return nil, err
  }
  keys = append(keys, k)
}
```

Reply:
Ah I see now, that makes sense. I've pushed a change which will only continue on `errNoCkaId` or `errNoPublicHalf` but will fail hard on any other errors.